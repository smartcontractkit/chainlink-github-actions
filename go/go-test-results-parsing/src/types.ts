import {z} from 'zod'

/**
 * The subschema for the test results that we want to handle
 * when generating the test result summary
 */
const handledTestStatuses = z.enum(['pass', 'fail'])
export type HandledTestResults = z.infer<typeof handledTestResultsSchema>
export const handledTestResultsSchema = z.object({
  Test: z.string().min(1).optional(),
  Action: handledTestStatuses,
  Elapsed: z.number().nonnegative()
})

/**
 * Possible test result actions formats
 *
 * Ideally, we would use `transform` to map the parsed test result to the
 * fields we want to use in the test result summary, but that breaks
 * discriminated unions.
 *
 * @see https://github.com/colinhacks/zod/issues/2315
 */
export type TestResult = z.infer<typeof TestResultSchema>
const TestResultSchema = z.discriminatedUnion('Action', [
  handledTestResultsSchema,
  z.object({
    Test: z.string().min(1).optional(),
    Action: z.literal('skip'),
    Elapsed: z.number().nonnegative()
  }),
  z.object({
    Test: z.string().min(1),
    Action: z.literal('run')
  }),
  z.object({
    Output: z.string().min(1),
    Action: z.literal('output'),
    Test: z.string().min(1).optional()
  }),
  z.object({
    Action: z.literal('start')
  }),
  z.object({
    Action: z.literal('pause')
  }),
  z.object({
    Action: z.literal('cont')
  })
])
/**
 * A representation of a file containing all test logs of
 * the JSONL format outputted by https://pkg.go.dev/cmd/test2json
 */
export const TestResultsSchema = z.array(TestResultSchema)

const outputMode = z.enum(['standard', 'json'])
export type OutputMode = z.infer<typeof outputMode>

const testRunFailures = z.object({
  TestsFailed: z.array(z.string()),
  PackageFailure: z.boolean()
})
export type TestRunFailures = z.infer<typeof testRunFailures>
