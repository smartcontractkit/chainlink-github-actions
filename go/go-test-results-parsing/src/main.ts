import * as core from '@actions/core'
import * as fs from 'fs'
import {KebabCasedProperties} from 'type-fest'
import {
  OutputMode,
  TestResult,
  TestResultsSchema,
  TestRunFailures
} from './types'

/**
 * Github action to parse the output of go test -json output to wanted formats with wanted data
 */

async function main(): Promise<void> {
  try {
    const resultsFile: string = getTypedInput('results-file')
    if (!resultsFile) {
      core.error('No results file provided')
    }
    const outputMode: OutputMode = getTypedInput('output-mode') as OutputMode
    const testResults: TestResult[] = deserializeTestResultsFile(resultsFile)
    const failedTests: TestRunFailures = getTestFailures(testResults)
    if (outputMode === 'standard') {
      standardTestOutput(failedTests.TestsFailed, testResults, true)
    } else if (outputMode === 'json') {
      jsonTestOutput(failedTests.TestsFailed, testResults, true)
    }
    if (failedTests.TestsFailed.length > 0) {
      throw new Error('Test Failures Found')
    } else if (
      failedTests.TestsFailed.length === 0 &&
      failedTests.PackageFailure
    ) {
      // we have a failure without a test failure, log out all the output
      logAllOutput(testResults)
      throw new Error('Test Failures Found')
    }
  } catch (error) {
    if (error instanceof Error) core.setFailed(error.message)
  }
}

/**
 * Deserialize the go test results file into a TestResult array
 * @param resultsFile The file with the -json go test results output
 * @returns The TestResult array with all the go test results data
 */
export function deserializeTestResultsFile(resultsFile: string): TestResult[] {
  const fileData = fs.readFileSync(resultsFile, 'utf8')
  // Split the content by newline characters
  const lines = fileData.split('\n').filter(line => line.trim() !== '')

  // Parse each line as a JSON object and collect the parsed objects in an array
  const jsonArray = lines.map(line => JSON.parse(line))
  return TestResultsSchema.parse(jsonArray)
}

/**
 * Get all the failed test names from the test results array
 * @param testResults The test results to filter through
 * @returns The failed test names
 */
export function getTestFailures(testResults: TestResult[]): TestRunFailures {
  const names: string[] = []
  let packageFailure = false
  testResults.filter(testResult => {
    if (testResult.Action === 'fail') {
      if (testResult.Test) {
        names.push(testResult.Test)
      } else {
        // it is possible to have a package failure without a test failure
        // so we are storing whether we have a package failure to determine
        // if we need to output all the output for triage later for panic cases
        packageFailure = true
      }
    }
  })
  return {TestsFailed: names, PackageFailure: packageFailure} as TestRunFailures
}

/**
 * Standard test output.
 * Only failures and just the output in the original format.
 * @param failedTestNames The names of the failed tests to output
 * @param testResults The test results to filter
 * @param shouldLogOutput Should we log the output to the workflow console?
 * @returns The filtered list of test failures outputs
 */
export function standardTestOutput(
  failedTestNames: string[],
  testResults: TestResult[],
  shouldLogOutput: boolean
): TestResult[] {
  let outputString = ''
  const filteredResults: TestResult[] = testResults.filter(testResult => {
    if (testResult.Action === 'output') {
      if (testResult.Test && failedTestNames.includes(testResult.Test)) {
        outputString += testResult.Output
        return testResult
      }
    }
  })
  outputHandler(outputString, shouldLogOutput)
  return filteredResults
}

/**
 * json test output.
 * Only failures and other output that doesn't inclue a test name.
 * Printed in the original broken json array format compatible with gotestfmt
 * @param failedTestNames
 * @param testResults
 * @param shouldLogOutput
 * @returns The filtered list of test result objects
 */
export function jsonTestOutput(
  failedTestNames: string[],
  testResults: TestResult[],
  shouldLogOutput: boolean
): TestResult[] {
  let outputString = ''
  const filteredResults: TestResult[] = testResults.filter(testResult => {
    if (
      testResult.Action === 'output' ||
      testResult.Action === 'run' ||
      testResult.Action === 'fail'
    ) {
      if (testResult.Test) {
        // if we have a test name then we only want tests in the failed test list
        if (failedTestNames.includes(testResult.Test)) {
          outputString += `${JSON.stringify(testResult)}\n`
          return testResult
        }
      } else {
        // if we don't have a test name then we want to just output whatever it is
        outputString += `${JSON.stringify(testResult)}\n`
        return testResult
      }
    } else if (testResult.Action === 'start') {
      // keep this to make gotestfmt happy
      outputString += `${JSON.stringify(testResult)}\n`
      return testResult
    }
  })
  outputHandler(outputString, shouldLogOutput)
  return filteredResults
}

/**
 * Write the output to the workflow console if wanted
 * Also write the output to a file if the filename was provided
 * @param output The string to output
 * @param shouldLog Whether we should log to the console
 */
export function outputHandler(output: string, shouldLog: boolean): void {
  if (shouldLog) {
    core.info(output)
  }
  const outputFile: string = getTypedInput('output-file', false)
  if (outputFile) {
    fs.writeFileSync(outputFile, output)
  }
}

/**
 * We have edge cases where we have a failure but no test failure
 * In this case we want to output all the output for triage since we
 * don't necessarily know what will be useful
 * @param testResults The test results to print outputs from
 */
export function logAllOutput(testResults: TestResult[]): void {
  let outputString = ''
  for (const testResult of testResults) {
    if (testResult.Action === 'output') {
      outputString += testResult.Output
    }
  }
  core.info(
    'We had an error in the test run but no specific test had a failure log, logging out everything for triage'
  )
  core.info(outputString)
}

/**
 * Takes kebob cased inputs and enforces they match the expected inputs
 * @param inputKey The input key to get
 * @param required Is the input required?
 * @returns The input value
 */
function getTypedInput(
  inputKey: keyof KebabCasedProperties<{
    resultsFile: never
    outputMode: never
    outputFile: never
  }>,
  required = true
): string {
  return core.getInput(inputKey, {required, trimWhitespace: true})
}

// Run the main function
main()
