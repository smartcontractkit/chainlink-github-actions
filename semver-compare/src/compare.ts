import semver from 'semver'

export type Operator = 'gt' | 'lt' | 'eq'

export function versionCompare(
  version1: string,
  operator: Operator,
  version2: string
): boolean {
  // Check for required input:
  if (!version1 || !version2 || !operator) {
    throw new Error('Required inputs not specified.')
  }

  // Validate the inputs:
  if (!semver.valid(version1) || !semver.valid(version2)) {
    throw new Error('Invalid version(s).')
  }

  if (!['gt', 'lt', 'eq'].includes(operator)) {
    throw new Error('Invalid operator.')
  }

  // Compare the versions:
  const result = semver[operator](version1, version2)
  return result
}
