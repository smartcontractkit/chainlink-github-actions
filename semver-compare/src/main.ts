import * as core from '@actions/core';
import semver from 'semver';

/**
 * Simple GitHub Actions wrapper around the semver library.
 */

async function run(): Promise<void> {
      try { 
        const version1 = core.getInput('version1');
        const operator = core.getInput('operator');
        const version2 = core.getInput('version2');
        core.debug(`Comparing ${version1} ${operator} ${version2}`);

        // Define valid operators
        type Operator = 'gt' | 'lt' | 'eq';

        // Check for required input:
        if (!version1 || !version2 || !operator) {
          throw new Error('Required inputs not specified.');
        }

        // Validate the inputs:
        if (!semver.valid(version1) || !semver.valid(version2)) {
          throw new Error('Invalid version(s).');
        }

        if (!['gt', 'lt', 'eq'].includes(operator as Operator)) {
          throw new Error('Invalid operator.');
        }

        // Compare the versions:
        const result = semver[operator as Operator](version1, version2);
        core.setOutput('result', result);
    } 
    catch (error) {
      if (error instanceof Error) core.setFailed(error.message);
    }
  }

  run()
  