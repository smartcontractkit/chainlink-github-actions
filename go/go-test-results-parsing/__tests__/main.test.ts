import {
  deserializeTestResultsFile,
  jsonTestOutput,
  getTestFailures,
  standardTestOutput
} from '../src/main'
import {TestResult} from '../src/types'
const file = './__tests__/fixtures/go_test_results_input.json'
describe('output testing', () => {
  test('can deserialize file', () => {
    const testResults: TestResult[] = deserializeTestResultsFile(file)
    expect(testResults).toMatchSnapshot()
    expect(testResults.length).toBe(11)
  })
  test('failed test list', () => {
    const testResults: TestResult[] = deserializeTestResultsFile(file)
    const failedTests: string[] = getTestFailures(testResults)
    expect(failedTests).toMatchSnapshot()
    expect(failedTests.length).toBe(1)
  })
  test('standard test output', () => {
    const testResults: TestResult[] = deserializeTestResultsFile(file)
    const failedTests: string[] = getTestFailures(testResults)
    const filteredResults: TestResult[] = standardTestOutput(
      failedTests,
      testResults,
      true
    )
    expect(filteredResults).toMatchSnapshot()
    expect(filteredResults.length).toBe(1)
  })
  test('json test output', () => {
    const testResults: TestResult[] = deserializeTestResultsFile(file)
    const failedTests: string[] = getTestFailures(testResults)
    const filteredResults: TestResult[] = jsonTestOutput(
      failedTests,
      testResults,
      true
    )
    expect(filteredResults).toMatchSnapshot()
    expect(filteredResults.length).toBe(5)
  })
})
