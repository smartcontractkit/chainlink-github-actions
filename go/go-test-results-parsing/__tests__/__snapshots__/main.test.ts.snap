// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`output testing can deserialize file 1`] = `
[
  {
    "Action": "start",
  },
  {
    "Action": "pause",
  },
  {
    "Action": "cont",
  },
  {
    "Action": "run",
    "Test": "TestOCIChart",
  },
  {
    "Action": "output",
    "Output": "=== RUN   TestOCIChart
",
    "Test": "TestOCIChart",
  },
  {
    "Action": "output",
    "Output": "=== RUN   Test5NodesPlus2MiningGethsReorgEnv
",
    "Test": "Test5NodesPlus2MiningGethsReorgEnv",
  },
  {
    "Action": "output",
    "Output": "blah blah blah
",
  },
  {
    "Action": "skip",
    "Elapsed": 1,
    "Test": "TestMinResources5NodesEnvWithBlockscoutSkip",
  },
  {
    "Action": "pass",
    "Elapsed": 506.4,
    "Test": "TestOCIChart",
  },
  {
    "Action": "fail",
    "Elapsed": 522.96,
    "Test": "Test5NodesPlus2MiningGethsReorgEnv",
  },
  {
    "Action": "fail",
    "Elapsed": 551.993,
  },
]
`;

exports[`output testing failed test list 1`] = `
[
  "Test5NodesPlus2MiningGethsReorgEnv",
]
`;

exports[`output testing json test output 1`] = `
[
  {
    "Action": "start",
  },
  {
    "Action": "output",
    "Output": "=== RUN   Test5NodesPlus2MiningGethsReorgEnv
",
    "Test": "Test5NodesPlus2MiningGethsReorgEnv",
  },
  {
    "Action": "output",
    "Output": "blah blah blah
",
  },
  {
    "Action": "fail",
    "Elapsed": 522.96,
    "Test": "Test5NodesPlus2MiningGethsReorgEnv",
  },
  {
    "Action": "fail",
    "Elapsed": 551.993,
  },
]
`;

exports[`output testing standard test output 1`] = `
[
  {
    "Action": "output",
    "Output": "=== RUN   Test5NodesPlus2MiningGethsReorgEnv
",
    "Test": "Test5NodesPlus2MiningGethsReorgEnv",
  },
]
`;
